### PR TITLE
New features

### PR DESCRIPTION
**Pull request by akelanova**
Add two new controllers. 
The first provides statistical information on the performance indicators in the country. It has 4 parameters:
- country's name;
- number of vacancies;
- number of employees;
- economic level.
The second controller provides statistical information about average indicators for developers in the countries. It returns 5 parameters:
- country's name;
- average salary;
- average age;
- average experience;
- average number of employees in companies.

### CODE REVIEW

File: src/main/java/com/spring/app/MessageEnum.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
     private String message;
 
     /**
-     * Complete the message template with the args value information
+     * Completes the message template with the args value information
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/MessageEnum.java:28]: няш

File: src/main/java/com/spring/app/domain/Country.java [hunk 2]
```diff
@@ -66,7 +66,45 @@
...
+                        query = "SELECT " +
+                                "       country_name_field       AS country, " +
+                                "       AVG(avgSalary)           AS salary," +
+                                "       AVG(avgAge)              AS age, " +
+                                "       AVG(avgExp)              AS experience, " +
+                                "       AVG(number_of_employees) AS number_of_employees " +
+                                "FROM (SELECT COUNT(country.id)                                                                          AS number_of_employees, " +
+                                "             country.name                                                                               AS country_name_field, " +
+                                "             AVG(vacancy.salary)                                                                        AS avgSalary, " +
+                                "             AVG((cast(extract(year FROM cast(:to_date AS DATE)) AS int)) - employee.birth_year)        AS avgAge, " +
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/Country.java]: Good, but let's make more performance-optimise. You can extract year at the top select statement. This will be calculated only once, when the inner select completes.

File: src/main/java/com/spring/app/domain/Country.java [hunk 3]
```diff
@@ -78,15 +116,17 @@
 public class Country extends BaseEntity {
     public static String MEAN_SALARY = "MEAN_SALARY";
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/Country.java]: all constants must be final

File: src/main/java/com/spring/app/dto/CountryProfileDTO.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Data transfer object for storage info about country's profile
+ *
+ * @author Lyubov Ruzanova
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class CountryProfileDTO implements Serializable {
+    private Long numberOfVacancies;
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/dto/CountryProfileDTO.java]: Some sacral sense is enclosed using number Wrapper instead of simple number?

File: src/main/java/com/spring/app/dto/EmployeeProfileDTO.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * @author Lyubov Ruzanova
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class EmployeeProfileDTO implements Serializable {
+    private Double salary;
+    private Integer age;
+    private Integer experience;
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/dto/EmployeeProfileDTO.java]: same

File: src/main/java/com/spring/app/services/CountryService.java [hunk 6]
```diff
@@ -41,43 +44,72 @@ public CountryService(@Autowired CountryDAO repository) {
...
-     * @param name - country's name
+     * Calculates the average employee's experience in the country
+     * @param countryName - country's Name
+     * @param date1       - start date of the period
+     * @param date2       - end date of the period
      * @return the Integer value of the average experience
      */
-    public Integer getAvgExperience(String name, Date date1, Date date2) {
-        return repository.findAvgExperience(name, date1, date2);
+    public Integer getAvgExperience(String countryName, Date date1, Date date2) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/CountryService.java]: Integer wrapper

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                    Long empNumber = LongResultMapper.getLong(object[EMP_NUMBER_INDEX]);
+
+                    CountryProfileDTO countryProfileDTO = new CountryProfileDTO();
+                    countryProfileDTO.setNumberOfEmployees(empNumber);
+                    countryProfileDTO.setNumberOfVacancies(vacNumber);
+                    countryProfileDTO.setEconomicLevel(economyIndex);
+
+                    dtoHashMap.put(name, countryProfileDTO);
+                } else {
+                    throw new DataNotFoundException("The received value is incorrect.");
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: Message in constants, do not throw any exceptions in mapper. Just return an empty collection

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    private static final int EMP_NUMBER_INDEX = 3;
+
+    /**
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, CountryProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            Map<String, CountryProfileDTO> dtoHashMap = new HashMap<>();
+            for (Object[] object : objects) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: stream?

File: src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java [hunk 9]
```diff
@@ -19,16 +19,27 @@
...
         throw new DataNotFoundException("Data not found.");
     }
+
+    /**
+     * Resolver for an object that should contains a double value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static Double getDouble(Object object) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java]: seems like a utility method. Extract it as a default interface method

File: src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * {@InheritDoc}
+     */
+    @Override
+    public EconomicLevel map(List<Object> objects) throws DataNotFoundException {
+        if (objects != null && !objects.isEmpty()) {
+            Object object = objects.get(SALARY_INDEX);
+            return getEconomicLevel(object);
+        }
+        throw new DataNotFoundException("Data not found.");
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java]: do not throw exceptions here

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Resolver for an object that should contains a String value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    private String resolveStringValue(Object object) {
+        if (object != null) {
+            if (object instanceof String) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: useless

File: src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import com.spring.app.EconomicLevel;
+
+import java.util.List;
+
+/**
+ * Maps data set to single value and cast it to EconomicLevel
+ *
+ * @author Lyubov Ruzanova
+ */
+public class EconomyLevelResultMapper implements ResultMapper<List<Object>, EconomicLevel> {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java]: is this really needed?

File: src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+     * Resolver for an object that should contains the economic level value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static EconomicLevel getEconomicLevel(Object object) {
+        if (object == null){
+            throw new DataNotFoundException("Data not found.");
+        }
+        if (object instanceof EconomicLevel) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java]: this will never work actually

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    private static final int NUM_OF_EMP_INDEX = 4;
+
+    /**
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, EmployeeProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            Map<String, EmployeeProfileDTO> dtoHashMap = new HashMap<>();
+            for (Object[] object : objects) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: no exceptions,stream

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        throw new DataNotFoundException("Data not found.");
+    }
+
+    /**
+     * Resolver for an object that should contains a String value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    private String resolveStringValue(Object object) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: code dublicate

File: src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java [hunk 16]
```diff
@@ -19,15 +20,28 @@
...
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static Integer getInteger(Object object) {
+        if (object == null) {
+            throw new DataNotFoundException("Data not found.");
+        }
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).intValue();
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java]: code dublicate

File: src/main/java/com/spring/app/services/mappers/LongResultMapper.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * Resolver for an object that should contains an Long value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static Long getLong(Object object) {
+        if (object == null){
+            throw new DataNotFoundException("Data not found.");
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/LongResultMapper.java]: code dublicate

**Review by EvgeniiRay — CHANGES_REQUESTED** [review]

File: src/main/java/com/spring/app/dto/CountryProfileDTO.java (hunk 4 above)
Comment (akelanova) [reply src/main/java/com/spring/app/dto/CountryProfileDTO.java]: now it does not matter) understood, I will use a primitive type

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java (hunk 7 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: Oh, really) You are right :) You mean there is no mistake, it's just no-result response from DB. I missed it. But if I find out a that result isn't instance of object that I want I will throw an error?

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java (hunk 8 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: stream, ok)

Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: :(

Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: (:

File: src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java (hunk 13 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java]: i agree with you

File: src/main/java/com/spring/app/services/CountryService.java (hunk 6 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/CountryService.java]: Changed it on primitive type

File: src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java (hunk 9 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java]: I have a solution. But I do not sure if this is the best decision

File: src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java (hunk 12 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/EconomyLevelResultMapper.java]: mmm... not really, I can relocate it as a simple method of the CountryProfileDTOResultMapper class

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java (hunk 11 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: what is wrong?

File: src/main/java/com/spring/app/services/mappers/LongResultMapper.java (hunk 17 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/LongResultMapper.java]: Can you explain how can I fix this?

File: src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java (hunk 16 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java]: I don't know how to resolve this

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java (hunk 11 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: I can combine this lines together and leave only instanceof ?)

Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: can i mark it resolved?

File: src/main/java/com/spring/app/domain/Country.java (hunk 2 above)
Comment (akelanova) [reply src/main/java/com/spring/app/domain/Country.java]: I did a new one

File: src/main/java/com/spring/app/domain/Country.java [hunk 18]
```diff
@@ -66,7 +66,82 @@
...
+                                "            country.name      AS country_name " +
+                                "     FROM country " +
+                                "             INNER JOIN company ON (country.id = company.country_id " +
+                                "                                      AND company.date_opened <= :to_date " +
+                                "                                      AND (company.date_closed ISNULL OR company.date_closed >= :from_date)) " +
+                                "             INNER JOIN vacancy ON (company.id = vacancy.company_id " +
+                                "                                      AND vacancy.opened = TRUE " +
+                                "                                      AND vacancy.date_opened >= :from_date " +
+                                "                                      AND (vacancy.date_closed ISNULL OR vacancy.date_closed >= :to_date)) " +
+                                "     GROUP BY company.name, country.name) AS companies_in_country_stat " +
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/Country.java]: о ма гад

File: src/main/java/com/spring/app/domain/Country.java (hunk 2 above)
Comment (EvgeniiRay) [reply src/main/java/com/spring/app/domain/Country.java]: @LetiLubov which actually performs N^2 times slower. Good job!

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Resolver for an value of the Object type that must contains a economic level value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public EconomicLevel resolveEconomicLevelValue(Object object) throws DataNotFoundException {
+        if (object instanceof String) {
+            try {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: useless

File: src/main/java/com/spring/app/dto/CountryProfileDTO.java [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+@Getter
+@Setter
+@NoArgsConstructor
+public class CountryProfileDTO implements Serializable {
+    private long numberOfVacancies;
+    private long numberOfEmployees;
+    private EconomicLevel economicLevel;
+
+    public CountryProfileDTO(long numberOfVacancies, long numberOfEmployees, EconomicLevel economicLevel) {
+        this.numberOfVacancies = numberOfVacancies;
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/dto/CountryProfileDTO.java]: @ all arg constructor

File: src/main/java/com/spring/app/dto/EmployeeProfileDTO.java [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+@NoArgsConstructor
+public class EmployeeProfileDTO implements Serializable {
+    private double salary;
+    private int age;
+    private int experience;
+    private int numberOfEmployees;
+
+    public EmployeeProfileDTO(double salary, int age, int experience, int numberOfEmployees) {
+        this.salary = salary;
+        this.age = age;
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/dto/EmployeeProfileDTO.java]: same

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, CountryProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            Map<String, CountryProfileDTO> dtoHashMap =
+                    objects.stream()
+                            .filter(x -> x != null)
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: Objects::nonNull

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+     * Resolver for an value of the Object type that must contains a economic level value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public EconomicLevel resolveEconomicLevelValue(Object object) throws DataNotFoundException {
+        if (object instanceof String) {
+            try {
+                return EconomicLevel.valueOf((String) object);
+            } catch (IllegalArgumentException ex) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: still not fixed

File: src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java [hunk 24]
```diff
@@ -19,16 +19,25 @@
...
+     * @return converted object
+     */
+    @Override
+    public Double convertObject(Object object) throws DataNotFoundException {
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).doubleValue();
+        } else if (object instanceof Double) {
+            return (Double) object;
+        }
+        throw new DataNotFoundException("The received value is incorrect.");
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java]: exception is not needed

File: src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java [hunk 25]
```diff
@@ -19,15 +20,26 @@
...
+     *
+     * @param object - input
+     * @return converted object
+     */
+    @Override
+    public Integer convertObject(Object object) throws DataNotFoundException {
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).intValue();
+        } else if (object instanceof BigInteger) {
+            return ((BigInteger) object).intValue();
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java]: code dublicate

File: src/main/java/com/spring/app/services/mappers/StringResolveUtils.java [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+public final class StringResolveUtils {
+    /**
+     * Resolver for an object that should contains a String value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static String resolveStringValue(Object object) {
+            if (object instanceof String) {
+                return object.toString();
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/StringResolveUtils.java]: very strange stuff

**Review by EvgeniiRay — CHANGES_REQUESTED** [review]

File: src/main/java/com/spring/app/domain/Country.java (hunk 18 above)
Comment (akelanova) [reply src/main/java/com/spring/app/domain/Country.java]: ??????

File: src/main/java/com/spring/app/domain/Country.java (hunk 2 above)
Comment (akelanova) [reply src/main/java/com/spring/app/domain/Country.java]: i did a new one!

Comment (akelanova) [reply src/main/java/com/spring/app/domain/Country.java]: you will like it

File: src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java (hunk 16 above)
Comment (akelanova) [reply src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java]: i want to delete all this simple result mappers

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java (hunk 22 above)
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: still possible npe, you have required field COUNTRY_NAME, theoretically, it can be null, you should also check it

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+        return new HashMap<>();
+    }
+
+    /**
+     * Resolver for an value of the Object type that must contains a economic level value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public EconomicLevel resolveEconomicLevelValue(Object object) throws DataNotFoundException {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: useless , useless...

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+        if (object instanceof String) {
+            try {
+                return EconomicLevel.valueOf((String) object);
+            } catch (IllegalArgumentException ex) {
+                throw new DataNotFoundException("The country level is undefined.");
+            }
+        }
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java:57]: still exceptions

File: src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java (hunk 24 above)
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/DoubleResultMapper.java]: return default 0d

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+    @Override
+    public Map<String, EmployeeProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            Map<String, EmployeeProfileDTO> dtoHashMap =
+                    objects.stream()
+                            .filter(x -> x!= null)
+                            .collect(Collectors.toMap(
+                                    object -> StringResolveUtils.resolveStringValue(object[COUNTRY_INDEX]),
+                                    object -> new EmployeeProfileDTO(
+                                            new DoubleResultMapper().convertObject(object[SALARY_INDEX]),
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: each entry you create a new object, , it causes a huge performance impact

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, EmployeeProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            Map<String, EmployeeProfileDTO> dtoHashMap =
+                    objects.stream()
+                            .filter(x -> x!= null)
+                            .collect(Collectors.toMap(
+                                    object -> StringResolveUtils.resolveStringValue(object[COUNTRY_INDEX]),
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: NPE and still useless enum  resolver

File: src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java [hunk 31]
```diff
@@ -19,15 +20,26 @@
...
+        } else if (object instanceof Double) {
+            return ((Double) object).intValue();
+        }
+        throw new DataNotFoundException("The received value is incorrect.");
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/IntegerResultMapper.java]: why are you throwing exception here?

File: src/main/java/com/spring/app/services/mappers/StringResolveUtils.java [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+import com.spring.app.DataNotFoundException;
+
+public final class StringResolveUtils {
+    /**
+     * Resolver for an object that should contains a String value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public static String resolveStringValue(Object object) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/StringResolveUtils.java]: useless

**Review by EvgeniiRay — CHANGES_REQUESTED** [review]

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 33]
```diff
@@ -18,72 +23,109 @@
...
      * @param date2       - end date of the period
      * @param countryName - country's name
      * @return the double value of the average salary
      */
-    public Double findMeanSalary(String countryName, Date date1, Date date2) {
-        return new DoubleResultMapper().map(em.createNamedQuery(Country.MEAN_SALARY)
+    public double findMeanSalary(String countryName, Date date1, Date date2) {
+        Double result = (Double) em.createNamedQuery(Country.MEAN_SALARY)
                 .setParameter("country_name", countryName)
                 .setParameter("from_date", date1)
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java:37]: result is not always double, big decimal possible

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 34]
```diff
@@ -18,72 +23,109 @@
...
                 .setParameter("to_date", date2)
-                .getResultList());
+                .getResultList()
+                .get(0);
+        return (result != null) ? result : 0d;
     }
 
     /**
-     * Create query to DB to find out the average number of employees in companies in the country in period of time
+     * Creates query to DB to finds out the average number of employees in companies in the country in period of time
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java]: bad english , Creates smth to do smth

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 35]
```diff
@@ -18,72 +23,109 @@
...
      *
      * @param date1       - start date of the period
      * @param date2       - end date of the period
      * @param countryName - country's name
      * @return the Integer value of the average number of employees
      */
-    public Integer findAvgNumberOfEmployees(String countryName, Date date1, Date date2) {
-        return new IntegerResultMapper().map(em.createNamedQuery(Country.AVG_NUMBER_OF_EMPLOYEES)
+    public int findAvgNumberOfEmployeesInCompanies(String countryName, Date date1, Date date2) {
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java:51]: same

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 36]
```diff
@@ -18,72 +23,109 @@
...
      * @param date1       - start date of the period
      * @param date2       - end date of the period
      * @param countryName - country's name
      * @return the Integer value of the average experience
      */
-    public Integer findAvgExperience(String countryName, Date date1, Date date2)  {
-        return new IntegerResultMapper().map(em.createNamedQuery(Country.AVG_EXPERIENCE)
+    public int findAvgExperience(String countryName, Date date1, Date date2) {
+        BigDecimal result = (BigDecimal) em.createNamedQuery(Country.AVG_EXPERIENCE)
                 .setParameter("country_name", countryName)
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java:70]: same

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 37]
```diff
@@ -18,72 +23,109 @@
...
      * @param countryName - country's name
      * @return the Integer value of the average experience
      */
-    public Integer findAvgAge(String countryName, Date date1, Date date2) {
-        return new IntegerResultMapper().map(em.createNamedQuery(Country.AVG_AGE)
+    public int findAvgAge(String countryName, Date date1, Date date2) {
+        BigDecimal result = (BigDecimal) em.createNamedQuery(Country.AVG_AGE)
                 .setParameter("country_name", countryName)
+                .setParameter("from_date", date1)
+                .setParameter("to_date", date2)
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java:89]: same

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 38]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public Map<String, CountryProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            return objects.stream()
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toMap(
+                            object -> (object[COUNTRY_INDEX] != null) ?
+                                    object[COUNTRY_INDEX].toString() : "undefined",
+                            object -> new CountryProfileDTO(
+                                    (object[VAC_NUMBER_INDEX] != null) ?
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: null is not possible here

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 39]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, CountryProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            return objects.stream()
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toMap(
+                            object -> (object[COUNTRY_INDEX] != null) ?
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java]: you shouldn't add such result to a collection. It should be filtered

File: src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java [hunk 40]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Resolver for an value of the Object type that must contains a economic level value
+     *
+     * @param object - input
+     * @return converted object
+     */
+    public EconomicLevel resolveEconomicLevelValue(Object object) {
+        try {
+            return EconomicLevel.valueOf((String) object);
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/CountryProfileDTOResultMapper.java:54]: good

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 41]
```diff
@@ -0,0 +1,49 @@
...
+     * {@InheritDoc}
+     */
+    @Override
+    public Map<String, EmployeeProfileDTO> map(List<Object[]> objects) {
+        if (objects != null && !objects.isEmpty()) {
+            return objects.stream()
+                            .filter(Objects::nonNull)
+                            .collect(Collectors.toMap(
+                                    object -> (object[COUNTRY_INDEX] != null) ?
+                                            object[COUNTRY_INDEX].toString() : "undefined",
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: same

File: src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java [hunk 42]
```diff
@@ -0,0 +1,49 @@
...
+                                    object -> new EmployeeProfileDTO(
+                                            (object[SALARY_INDEX] != null) ?
+                                                    (Double) object[SALARY_INDEX] : 0d,
+                                            (object[AGE_INDEX] != null) ?
+                                                    ((BigDecimal) object[AGE_INDEX]).intValue() : 0,
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/services/mappers/EmployeeProfileDTOResultMapper.java]: same

**Review by EvgeniiRay — CHANGES_REQUESTED** [review]

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java (hunk 34 above)
Comment (akelanova) [reply src/main/java/com/spring/app/domain/dao/CountryDAO.java]: what do you mean?

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java (hunk 37 above)
Comment (akelanova) [reply src/main/java/com/spring/app/domain/dao/CountryDAO.java:89]: Do you want to add tests on null?

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java (hunk 34 above)
Comment (EvgeniiRay) [reply src/main/java/com/spring/app/domain/dao/CountryDAO.java]: > what do you mean?

Creates query to DB to **FIND**

File: src/main/java/com/spring/app/domain/dao/CountryDAO.java [hunk 43]
```diff
@@ -18,72 +24,105 @@
...
                 .setParameter("from_date", date1)
                 .setParameter("to_date", date2)
-                .getResultList());
+                .getResultList()
+                .get(0));
     }
 
     /**
-     * Create query to DB to find out the average age of employees in the country in period of time
+     * Creates query to DB to finds out the average age of employees in the country in period of time
```
Comment (EvgeniiRay) [inline src/main/java/com/spring/app/domain/dao/CountryDAO.java]: same

**Review by EvgeniiRay — CHANGES_REQUESTED** [review]

**Review by EvgeniiRay — APPROVED** [review]
